How to render HTML pages as an HTTP server response Using Node.js (#1742)
section-engineering-education/engineering-education: articles/rendering-html-pages-as-a-http-server-response-using-node-js/index.md

When developing web applications, you might need to render the HTML components inside your server. This helps you create interactive pages on the client side once a request to access these pages is made. There are various ways of hosting your HTML pages (website), for example;

- Rendering your client-side on its own using frameworks for such as React or,

> - Rend the pages directly from the server - when the browser tries to access a route specified in your server, the server will load these HTML pages upon users' requests.

Reviewer: Break this into 2-3 simpler sentences.

This guide will teach you how to render HTML elements and HTML pages on your server using Node.js.

### Prerequisites

Basic knowledge of using [Node.js](https://www.youtube.com/watch?v=fBNz5xF-Kx4) and [Express.js](/engineering-education/express/)

### What we will cover

- [Prerequisites](#prerequisites)
- [What we will cover](#what-we-will-cover)
- [Setting up](#setting-up)
- [Rendering inline Html elements as HTTP responses](#rendering-inline-html-elements-as-http-responses)
- [Render HTML web pages as server responses](#render-html-web-pages-as-server-responses)
- [Parsering form data to the server using the HTML forms](#parsering-form-data-to-the-server-using-the-html-forms)
- [Conclusion](#conclusion)

### Setting up

- [Download Node.js](https://nodejs.org/en/) and install it. Run `node -v` to test if the installation was successful.

```bash
$ node -v
v12.18.3 ## installed node.js version
```

> - Create a Node.js project/directory. Initialize the project within the directory you have created. Use `npm init -y` to automatically initialize this Node.js project or check this [guide](/engineering-education/beginner-guide-to-npm/) to learn more about using NPM.

Reviewer: Mention installation of npm in the setting up section.

```js
const express = require("express");
const app = express();

app.listen(3000, () => {
  console.log("Application started and Listening on port 3000");
});

app.get("/", (req, res) => {
  res.send("Hello world!");
});
```

>Run `nodemon` to start the server.

Whenever the server is running, accessing the route `http://localhost:3000/`, will give you plain text `hello world!`.

We can use the same server to render HTML elements as the server response instead of sending plain text.

Here is a list of some HTML elements. We can render them directly into our server by specifying the response to send when the default route is accessed.

```js
const express = require("express");
const app = express();

app.listen(3000, () => {
  console.log("Application started and Listening on port 3000");
});

app.get("/", (req, res) => {
  res.send("<html> <head>server Response</head><body><h1> This page was render direcly from the server <p>Hello there welcome to my website</p></h1></body></html>");
});
```

Restart the server and open route `http://localhost:3000/` on the browser.

> ![](server-rendered-html-elements.jpg)

Reviewer: Make sure to add text related to the content of the image. Also, modify all the image URLs to the following.
```![alt text](/engineering-education//rendering-html-pages-as-a-http-server-response-using-node-js/server-rendered-html-elements.jpg)```

`res.send` is sending individual bits of HTML data to the server, but if we want to send an entire web page such as an `index.html`, we have to use something different.

### Render HTML web pages as server responses

The above can be very tiresome and not the type of code you want to write inside your server. In a normal web page, HTML elements are written in a `.html` file. This way, you can write all your HTML elements, including your CSS styling, to layout these elements. This way, you separate the server files and the HTML elements, thus a clean code setup.

> To render an HTML file into the server using Express, we use `res.sendFile()`. This read and render the data that you include in your HTML files. This transfers the file over to the browser whenever it makes a [Get request](https://en.wikipedia.org/wiki/Hypertext_Transfer_Protocol#Request_methods) to the server. The server submits a response status with HTML rendered web content as the message body.

Reviewer: ```suggestion
To render an HTML file into the server using Express, we use `res.sendFile()`. This reads and renders the data included in one's HTML files. This transfers the file to the browser upon a [GET request](https://en.wikipedia.org/wiki/Hypertext_Transfer_Protocol#Request_methods) to the server. The server submits a response status with HTML rendered web content as the message body.
```

});

app.get("/", (req, res) => {
  console.log(__dirname)
});
```

Open the route `http://localhost:3000/` on the browser. This will print the values of `__dirname` into the console.

![](console-log-dirname.jpg)

As you can see, it gives you the exact path to reach your server location.

Now we can send HTML files to the server using `__dirname`.

To start with, I have created an HTML form (`index.html`) and included some CSS styling (`app.css`), as shown below.

```html
<!DOCTYPE html>
<html lang="en"> 
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <link rel="stylesheet" href="app.css">
  <link rel="stylesheet"
  href="//netdna.bootstrapcdn.com/twitter-bootstrap/2.3.2/css/bootstrap-combined.no-icons.min.css">
  <title>html form</title>
</head>
<body>
<div class="subscribe-container">
  <form method="POST">
    <input type="email" name="youremail" placeholder="Email Address" required>
    <input type="text" name="yourname" placeholder="Name" required>
    <input class="subscribe-button" type="submit" value="Subscribe">
  </form>
</div>
</body>
</html>
```

> `app.css`

Reviewer: ```suggestion
**File name**: _app.css_
```

Reviewer: Make the exact change everywhere file names are mentioned.

```css
.subscribe-container {
    max-width: 800px;
    margin: 60px auto;
    background-color: rgba(130, 184, 219, 0.5);
    border: 5px solid rgb(98, 143, 228);
}
.subscribe-container form {
    display: flex;
    flex-wrap: wrap;
}
.subscribe-container form input {
    margin: 15px;
    border: 1px solid rgb(98, 143, 228);
    padding: 0.4rem;
}
.subscribe-container form input {
    flex: 1 1 200px;
}

.subscribe-container form input[type="submit"] {
    border-radius: 3px;
    background-color: rgba(17, 228, 10, 0.5);
color: rgb(98, 143, 228);
}

.subscribe-container form input[type="email"] {
    flex: 1 1 250px;
}
```

Let us render the file into the server.

```js
const express = require("express");
const app = express();

app.listen(3000, () => {
  console.log("Application started and Listening on port 3000");
});

app.get("/", (req, res) => {
  res.sendFile(__dirname + "/index.html");
});
```

![](html-page-as-server-response.jpg)

> However, this did not load the CSS styling. `app.css` is a static file. To [static server files](https://expressjs.com/en/starter/static-files.html) included in the `index.html` use `express.static` as shown in the example below;

Reviewer: I think the second sentence is incomplete. Check that. It should be, to load static server files?

```js
const express = require("express");
const app = express();

app.listen(3000, () => {
  console.log("Application started and Listening on port 3000");
});

// serve your css as static
app.use(express.static(__dirname));

app.get("/", (req, res) => {
  res.sendFile(__dirname + "/index.html");
});
```

Save the file and open `http://localhost:3000/` in the browser, and the server will send a web page as expected.

![](html-page-with-css-as-server-response.jpg)

### Parsering form data to the server using the HTML forms

> The server is running. It is now returning HTML form as a response to the client (browser). Whenever you access this server's route, the `Get` request will be executed from the browser. However, what would happen if you fill this form with the data and press the button _subscribe_?

Reviewer: Maintain consistency while mentioning commands such as GET, POST, etc.
